lib/rack/request.rb

```diff
@@ -232,52 +278,87 @@ def xhr?
         get_header("HTTP_X_REQUESTED_WITH") == "XMLHttpRequest"
       end
 
-      def host_with_port
-        port = self.port
-        if port.nil? || port == DEFAULT_PORTS[scheme]
-          host
+      # The `HTTP_HOST` header.
+      def host_authority
+        get_header(HTTP_HOST)
+      end
+
+      def host_with_port(authority = self.authority)
+        host, address, port = split_authority(authority)
+
+        if port == DEFAULT_PORTS[self.scheme]
+          return host
         else
-          host = self.host
-          # If host is IPv6
-          host = "[#{host}]" if host.include?(':')
-          "#{host}:#{port}"
+          return authority
         end
       end
 
+      # Returns a formatted host, suitable for being used in a URI.
       def host
-        # Remove port number.
-        strip_port hostname.to_s
+        host, address, port = split_authority(self.authority)
+
+        return host
+      end
+
+      # Returns an address suitable for being used with `getaddrinfo`.
+      def hostname
+        host, address, port = split_authority(self.authority)
+
+        return address
       end
```

Comment:
This looks like inconsistency: the method `hostname` returns `address`… why? I think we should name it either `hostname` or `address` ecerywhere.

Reply:
It matches the implementation of `URI`.

Reply:
I mean, if you even just rename local variable to `hostname` — it'd be better. I don't see `address` as URI method or in its sources.

Reply:
Understood. It is actually confusing to me too in a way.

```
u = URI("http://[::1]/bar")
p u.hostname      #=> "::1"
p u.host          #=> "[::1]"
```

What does make sense to me is when you say "host_with_port". That means, the "escaped" host and `:port` appended. You can't have `hostname_with_port`. It's simply not possible because it becomes ambiguous to parse.

So, there's that logic... But yeah, I also see it's confusing.

Reply:
Yes, we could rename it `hostname` everywhere. I see what you are saying, and I don't think it would be a problem. Except that technically, I don't even know if `192.168.1.1` is a hostname? Or `1::` is a hostname? To me, something like `www.google.com` is a hostname. Not sure what is the right word/definition here.

Reply:
> Except that technically, I don't even know if `192.168.1.1` is a hostname? Or `1::` is a hostname? To me, something like `www.google.com` is a hostname. Not sure what is the right word/definition here.

I found [Wikipedia article](https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_hostnames):

> Hostnames are composed of a series of labels concatenated with dots. For example, "en.wikipedia.org" is a hostname. Each label must be from 1 to 63 characters long, and the entire hostname including the delimiting dots, but not a trailing dot, has a maximum of 253 ASCII characters.

So… no, IP addresses are not hostnames. Maybe we should write better code, logical, instead of blind following of URI (and, probably, we should improve URI).

Reply:
But "www.google.com" is also not an address.. is it?

Reply:
One part that I wondered about... we are returning lots of bits... ports, hosts... when I think that most users probably just want a rich URI object. Then the interface for the specific methods is less important/out of our hands.

e.g.

```
class Request
  def uri
    URI.parse("#{scheme}://#{authority}/#{path}")
  end
end
```

Reply:
> But "www.google.com" is also not an address.. is it?

I think, "address" is more common word, like address of a house. And in [Domain name article](https://en.wikipedia.org/wiki/Domain_name) I saw:

> Domain names are used in various networking contexts and for application-specific naming and **addressing** purposes.

But I also found [Network address article](https://en.wikipedia.org/wiki/Network_address), which is obviously more specific than common address, and there is no domain or hostname mentions inside, just numbers, IPs, MAC addresses…

So, that's not trivial question.

In such case we can follow stdlib, I think.

But, again, I've posted the first comment just about local variable name. :grimacing:

Reply:
So, should we change `def hostname` to `def address`?

Reply:
I know it's about local variable name but the entire thing is kind of confusing, so we should try to get it right for our users or we just perpetuate the chaos.

Reply:
So we have several options:

1/ Rename local variable `address` -> `hostname` and `return address` to `return hostname`.
2/ Rename method `def hostname` to `def address`.
3/ Remove `def hostname`.
4/ Add `def uri` which returns an instance of URI which represents what the URI was used to make the request.

Reply:
1 is like compliance with URI from stdlib.
2 is like try to improve naming, but, as I see, still with logic violations.
3 is radical, and, I guess, we still need for a helper method. If not — the best code is no code.
4 seems interesting, and "URI" sounds common enough for IPs and domains, especially with `URI` instance inside.

So, I can't say before trying, but I'd try to make in this order: 3, 4, 1, 2.

Reply:
I'll wait for feedback from @jeremyevans before I make any further changes.

Reply:
Just for the record, there are several new methods, and that includes `def hostname`. It's not released in Rack 2.1.